### ISSUE TITLE
[bug] `sg scan -U` fails because of leading and trailing empty lines

### ISSUE THREAD

**Issue by etiennebacher** [ast-grep/ast-grep#668]
Hello, this is a follow-up on #580, thanks a lot for having implemented it!

# ⏯ Playground Link

The issue doesn't appear in the playground: [link](https://ast-grep.github.io/playground.html#eyJtb2RlIjoiQ29uZmlnIiwibGFuZyI6InB5dGhvbiIsInF1ZXJ5IjoiIiwicmV3cml0ZSI6ImxvZ2dlci5sb2coJE1BVENIKSIsImNvbmZpZyI6ImlkOiBjaGFuZ2VfZGVmXG5sYW5ndWFnZTogUHl0aG9uXG5ydWxlOlxuICBwYXR0ZXJuOiB8XG4gICAgZGVmIGZvbygkWCk6XG4gICAgICAkJCRTXG5maXg6IHwtXG4gIGRlZiBiYXooJFgpOlxuICAgICQkJFNcblxuLS0tICMgdGhpcyBpcyBZQU1MIGRvYyBzZXBhcmF0b3IgdG8gaGF2ZSBtdWx0aXBsZSBydWxlcyBpbiBvbmUgZmlsZVxuXG5pZDogY2hhbmdlX3BhcmFtXG5ydWxlOlxuICBwYXR0ZXJuOiBmb28oJFgpXG5maXg6IGJheigkWCkiLCJzb3VyY2UiOiJcbmRlZiBmb28oeCk6XG4gICAgcmV0dXJuIHggKyAxXG5cbnkgPSBmb28oMilcbnByaW50KHkpIn0=)

# 💻 Code

This uses the same code and rules as in [this section](https://ast-grep.github.io/guide/rewrite-code.html#using-fix-in-yaml-rule) of the docs.

Code (it is important that there's an empty line at the beginning and at the end):

```python

def foo(x):
    return x + 1

y = foo(2)
print(y)

```

Pattern:

```yml
id: change_def
language: Python
rule:
  pattern: |
    def foo($X):
      $$$S
fix: |-
  def baz($X):
    $$$S

--- # this is YAML doc separator to have multiple rules in one file

id: change_param
rule:
  pattern: foo($X)
fix: baz($X)
```

# 🙁 Actual behavior

Running `sg scan -U` deletes the leading and trailing empty lines and returns this:
```python
dedef baz(x):
  return x + 1

y = fobaz(2)print(y)

```

# 🙂 Expected behavior

The result should be as in the docs and shouldn't depend on whether there are empty lines at the beginning or at the end.

**Comment by HerringtonDarkholme** [comment]
Thanks for the reporting!
However, I followed this on ast-grep 0.12.4 with the command `sg scan -U -r change.yml test.py`.

I cannot reproduce the issue. Would you like to use the `sg scan -r xxx.yml` command? Thanks

**Comment by etiennebacher** [comment]
I ran `sg scan -U -r xxx.yml` and `sg scan -U -r xxx.yml test.py` and the same bug appears in both cases. also have ast-grep 0.12.4.

However, removing `-U` shows that the detection is fine:
```
> sg scan -r inst/rules/rules.yml   
.\test.py
help[change_def]: 
@@ -0,6 +0,6 @@
1 1│ 
2  │-def foo(x):
3  │-    return x + 1
  2│+def baz(x):
  3│+  return x + 1
4 4│ 
5 5│ 
6 6│ y = foo(2)
.\test.py
help[change_param]: 
@@ -2,5 +2,5 @@
3 3│     return x + 1
4 4│ 
5 5│ 
6  │-y = foo(2)
  6│+y = baz(2)
7 7│ print(y)
```

**Comment by HerringtonDarkholme** [comment]
I still cannot reproduce it.

![image](https://github.com/ast-grep/ast-grep/assets/2883231/263cefb7-1b84-4bb3-9236-67145db38408)


Are you on windows?

**Comment by etiennebacher** [comment]
I am on windows but in your screenshot I don't see any empty line at the beginning of the python file. The bug appears when there are empty lines at the beginning and the end of the file.

**Comment by HerringtonDarkholme** [comment]
Oh thanks!  I can reproduce it!

**Comment by HerringtonDarkholme** [comment]
Thanks! I have identified the problem. Will fix it soon!
The problem is that leading whitespaces are not considered as source code in AST node. 
But the AST offset does count these whitespaces. The difference in offset makes the wrong string.

**Comment by HerringtonDarkholme** [comment]
Shipped in 0.12.5!

**Comment by etiennebacher** [comment]
Awesome, thanks a lot for the quick fix!
